arpack: use openblas on darwin; fix build; enable eigen

paparodeo opened the pull request:
change 07cdea2ae5005c6396eafd9d87e1cdec6c0c2a4b uncovered a misconfiguration in the darwin build resulting in some tests failing / segfaulting. Switch to using openblas which has no issues with the tests. Additionally, enable eigen as it was part of the build inputs but was unused as `EIGEN=ON` was not set in the build flags. Finally, added an option to use Accelerate building with the flags `-ff2c -fno-second-underscore` found https://github.com/opencollab/arpack-ng/blob/804fa3149a0f773064198a8e883bd021832157ca/.github/workflows/jobs.yml#L184-L192.

FWIW: homebrew also uses openblas. https://github.com/Homebrew/homebrew-core/blob/master/Formula/a/arpack.rb


https://github.com/NixOS/nixpkgs/pull/357259
https://hydra.nixos.org/build/282505341


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

paparodeo committed:
arpack: use openblas on darwin

Change 07cdea2ae5005c6396eafd9d87e1cdec6c0c2a4b uncovered a bug in how
we build arpack on darwin with the Accelerate. Disable the Accelerate
framework and use openblas the default nixpkgs blas/lapack
implementation.

paparodeo commented:
FWIW recent single precision darwin failures also with https://github.com/NixOS/nixpkgs/pull/367110 which is fixed using openblas.

paparodeo commented:
@ofborg build arpack

paparodeo commented:
cc: @fedeinthemix @SuperSandro2000

fedeinthemix commented:
I'm not familiar with the Darwin platform and don't have access to it. Therefore, I can't provide meaningful comments. However, the information at https://developer.apple.com/documentation/accelerate/blas/ may be relevant.

paparodeo commented:
> I'm not familiar with the Darwin platform and don't have access to it. Therefore, I can't provide meaningful comments. However, the information at https://developer.apple.com/documentation/accelerate/blas/ may be relevant.

thanks. we've had this issue with other packages and the solution is to use the nixpkgs default (openblas). mostly cc'd cause this change either needs to move forward or the other one backed out given that darwin is currently broken.

emilazy reviewed (approved):
Sure. (Is Accelerate.framework not handling single precision a known thing and definitely not something we’re holding wrong somehow?)

paparodeo commented:
> Sure. (Is Accelerate.framework not handling single precision a known thing and definitely not something we’re holding wrong somehow?)

- was also fixed here: https://github.com/NixOS/nixpkgs/pull/367110
- upstream issue https://github.com/xiaoyeli/superlu/issues/155

paparodeo commented:
> Sure. (Is Accelerate.framework not handling single precision a known thing and definitely not something we’re holding wrong somehow?)

tho they seem to make it work: https://github.com/opencollab/arpack-ng/actions/runs/11434914625/job/31809312118

https://github.com/opencollab/arpack-ng/blob/804fa3149a0f773064198a8e883bd021832157ca/.github/workflows/jobs.yml#L184-L192

the ` FFLAGS="-ff2c -fno-second-underscore"` seem to have something to do with making it not-segfault I think with accelerate. and `EXAMPLES=on` gets more tests however still fails 2 tests.

```
         83 - issue401_tst (Failed)
         84 - issue215_tst (Failed)
```

but using openblas (and without the `FFLAGS`) it fails the same tests.

paparodeo commented:
> the ` FFLAGS="-ff2c -fno-second-underscore"` seem to have something to do with making it not-segfault I think with accelerate. and `EXAMPLES=on` gets more tests however still fails 2 tests.
> 
> ```
>          83 - issue401_tst (Failed)
>          84 - issue215_tst (Failed)
>```

ok. so those tests fail on all platforms when `EIGEN=ON` which is what their test runners use but we don't set but i think we might've wanted to set as that uses the eigen library which is in the build inputs. `-ff2c` is required for `Accelerate` to not crash.  So when `FFLAGS=-ff2c` and `EIGEN=OFF` (the default) the tests all pass with Accelerate. So that's good.

I suppose now to figure out why when `EIGEN=ON` the above tests fail on all platforms (linux & darwin)

paparodeo commented:
> ```
>          83 - issue401_tst (Failed)
>          84 - issue215_tst (Failed)
> ```

ok -- `enableParallelChecking = false;` fixes these.

paparodeo committed:
arpack: minor cleanups

use ninja, cmakeBool and remove unneeded install_name_tool usage in
postFixup as the library name is already properly set by the build

paparodeo committed:
arpack: Enable eigenvalue-problems solver based on ICB and eigen

The arpack package included the eigen library and added it to the build
inputs but neglected to enable its use by setting `EIGEN=ON` in the
build flags.  Enable support for eigenvalue-problems solver based on ICB
and eigen and disable parallel checking as the tests fail when run in
parallel.

paparodeo committed:
arpack: add option to use macOS Accelerate

Copy darwin flags `-ff2c -fno-second-underscore` from workflow to use
Accelerate without segfaulting / failing tests and enable the Accelerate
framework when `useAccel` is true.
https://github.com/opencollab/arpack-ng/blob/804fa3149a0f773064198a8e883bd021832157ca/.github/workflows/jobs.yml#L184-L192

paparodeo commented:
OK. so updated the PR to now enable `EIGEN=ON` and parallel tests are disabled. Darwin still defaults to openblas with an option to enable Accelerate. Tests pass with ether option.

paparodeo commented:
@ofborg build arpack.passthru.tests arpack

paparodeo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 367594`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 9 packages marked as broken and skipped:</summary>
  <ul>
    <li>octavePackages.fem-fenics</li>
    <li>octavePackages.level-set</li>
    <li>octavePackages.parallel</li>
    <li>octavePackages.sparsersb</li>
    <li>octavePackages.tisean</li>
    <li>octavePackages.vibes</li>
    <li>octavePackages.vrml</li>
    <li>python312Packages.textnets</li>
    <li>python312Packages.textnets.dist</li>
  </ul>
</details>
<details>
  <summary>:x: 23 packages failed to build:</summary>
  <ul>
    <li>hal-hardware-analyzer</li>
    <li>librsb</li>
    <li>octavePackages.bim</li>
    <li>octavePackages.fits</li>
    <li>octavePackages.msh</li>
    <li>python312Packages.skytemple-ssb-debugger</li>
    <li>python312Packages.skytemple-ssb-debugger.dist</li>
    <li>python313Packages.explorerscript</li>
    <li>python313Packages.explorerscript.dist</li>
    <li>python313Packages.igraph</li>
    <li>python313Packages.igraph.dist</li>
    <li>python313Packages.kmapper</li>
    <li>python313Packages.kmapper.dist</li>
    <li>python313Packages.leidenalg</li>
    <li>python313Packages.leidenalg.dist</li>
    <li>python313Packages.skytemple-dtef</li>
    <li>python313Packages.skytemple-dtef.dist</li>
    <li>python313Packages.skytemple-files</li>
    <li>python313Packages.skytemple-files.dist</li>
    <li>python313Packages.skytemple-ssb-debugger</li>
    <li>python313Packages.skytemple-ssb-debugger.dist</li>
    <li>skytemple</li>
    <li>skytemple.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 97 packages built:</summary>
  <ul>
    <li>LPCNet</li>
    <li>arpack</li>
    <li>arpack-mpi</li>
    <li>calculix-ccx</li>
    <li>checkov</li>
    <li>checkov.dist</li>
    <li>flattenReferencesGraph</li>
    <li>flattenReferencesGraph.dist</li>
    <li>gama</li>
    <li>igraph</li>
    <li>igraph.dev</li>
    <li>igraph.doc</li>
    <li>jupyter-all</li>
    <li>libleidenalg</li>
    <li>minc_widgets</li>
    <li>octave</li>
    <li>octaveFull</li>
    <li>octavePackages.arduino</li>
    <li>octavePackages.audio</li>
    <li>octavePackages.bsltl</li>
    <li>octavePackages.cgi</li>
    <li>octavePackages.communications</li>
    <li>octavePackages.control</li>
    <li>octavePackages.data-smoothing</li>
    <li>octavePackages.database</li>
    <li>octavePackages.dataframe</li>
    <li>octavePackages.dicom</li>
    <li>octavePackages.divand</li>
    <li>octavePackages.doctest</li>
    <li>octavePackages.econometrics</li>
    <li>octavePackages.financial</li>
    <li>octavePackages.fpl</li>
    <li>octavePackages.fuzzy-logic-toolkit</li>
    <li>octavePackages.ga</li>
    <li>octavePackages.general</li>
    <li>octavePackages.generate_html</li>
    <li>octavePackages.geometry</li>
    <li>octavePackages.gsl</li>
    <li>octavePackages.image</li>
    <li>octavePackages.image-acquisition</li>
    <li>octavePackages.instrument-control</li>
    <li>octavePackages.interval</li>
    <li>octavePackages.io</li>
    <li>octavePackages.linear-algebra</li>
    <li>octavePackages.lssa</li>
    <li>octavePackages.ltfat</li>
    <li>octavePackages.mapping</li>
    <li>octavePackages.matgeom</li>
    <li>octavePackages.miscellaneous</li>
    <li>octavePackages.mvn</li>
    <li>octavePackages.nan</li>
    <li>octavePackages.ncarray</li>
    <li>octavePackages.netcdf</li>
    <li>octavePackages.nurbs</li>
    <li>octavePackages.ocl</li>
    <li>octavePackages.octclip</li>
    <li>octavePackages.octproj</li>
    <li>octavePackages.optics</li>
    <li>octavePackages.optim</li>
    <li>octavePackages.optiminterp</li>
    <li>octavePackages.quaternion</li>
    <li>octavePackages.queueing</li>
    <li>octavePackages.signal</li>
    <li>octavePackages.sockets</li>
    <li>octavePackages.splines</li>
    <li>octavePackages.statistics</li>
    <li>octavePackages.stk</li>
    <li>octavePackages.strings</li>
    <li>octavePackages.struct</li>
    <li>octavePackages.symbolic</li>
    <li>octavePackages.tsa</li>
    <li>octavePackages.video</li>
    <li>octavePackages.windows</li>
    <li>octavePackages.zeromq</li>
    <li>octopus</li>
    <li>openems</li>
    <li>python312Packages.explorerscript</li>
    <li>python312Packages.explorerscript.dist</li>
    <li>python312Packages.igraph</li>
    <li>python312Packages.igraph.dist</li>
    <li>python312Packages.kmapper</li>
    <li>python312Packages.kmapper.dist</li>
    <li>python312Packages.leidenalg</li>
    <li>python312Packages.leidenalg.dist</li>
    <li>python312Packages.python-csxcad</li>
    <li>python312Packages.python-csxcad.dist</li>
    <li>python312Packages.python-openems</li>
    <li>python312Packages.python-openems.dist</li>
    <li>python312Packages.scikit-tda</li>
    <li>python312Packages.scikit-tda.dist</li>
    <li>python312Packages.skytemple-dtef</li>
    <li>python312Packages.skytemple-dtef.dist</li>
    <li>python312Packages.skytemple-files</li>
    <li>python312Packages.skytemple-files.dist</li>
    <li>python313Packages.python-csxcad</li>
    <li>python313Packages.python-csxcad.dist</li>
    <li>vpv</li>
  </ul>
</details>

<details><summary>all failures currently broken on master</summary>

```
Build Status for hal-hardware-analyzer.x86_64-linux on master
✖ (Failed) hal-hardware-analyzer-4.4.1 from 2024-12-23 - https://hydra.nixos.org/build/282858541
Build Status for librsb.x86_64-linux on master
✖ (Failed) librsb-1.2.0.10 from 2024-12-23 - https://hydra.nixos.org/build/282862544
Build Status for octavePackages.bim.x86_64-linux on master
✖ (Dependency failed) octave-9.3.0-bim-1.1.6 from 2024-12-23 - https://hydra.nixos.org/build/282827223
Build Status for octavePackages.fits.x86_64-linux on master
✖ (Failed) octave-9.3.0-fits-1.0.7 from 2024-12-23 - https://hydra.nixos.org/build/282836762
Build Status for octavePackages.msh.x86_64-linux on master
✖ (Dependency failed) octave-9.3.0-msh-1.0.12 from 2024-12-23 - https://hydra.nixos.org/build/282906993
Build Status for python312Packages.skytemple-ssb-debugger.x86_64-linux on master
✖ (Dependency failed) python3.12-skytemple-ssb-debugger-1.8.2 from 2024-12-23 - https://hydra.nixos.org/build/282733000
Build Status for python313Packages.explorerscript.x86_64-linux on master
✖ (Dependency failed) python3.13-explorerscript-0.2.1.post2 from 2024-12-23 - https://hydra.nixos.org/build/282874744
Build Status for python313Packages.igraph.x86_64-linux on master
✖ (Dependency failed) python3.13-igraph-0.11.8 from 2024-12-23 - https://hydra.nixos.org/build/282752950
Build Status for python313Packages.kmapper.x86_64-linux on master
✖ (Dependency failed) python3.13-kmapper-2.1.0 from 2024-12-23 - https://hydra.nixos.org/build/282865902
Build Status for python313Packages.leidenalg.x86_64-linux on master
✖ (Dependency failed) python3.13-leidenalg-0.10.2 from 2024-12-23 - https://hydra.nixos.org/build/282892400
Build Status for python313Packages.skytemple-dtef.x86_64-linux on master
✖ (Dependency failed) python3.13-skytemple-dtef-1.6.1 from 2024-12-23 - https://hydra.nixos.org/build/282834043
Build Status for python313Packages.skytemple-files.x86_64-linux on master
✖ (Dependency failed) python3.13-skytemple-files-1.8.3 from 2024-12-23 - https://hydra.nixos.org/build/282735686
Build Status for python313Packages.skytemple-ssb-debugger.x86_64-linux on master
✖ (Dependency failed) python3.13-skytemple-ssb-debugger-1.8.2 from 2024-12-23 - https://hydra.nixos.org/build/282819748
Build Status for skytemple.x86_64-linux on master
✖ (Dependency failed) skytemple-1.8.3 from 2024-12-23 - https://hydra.nixos.org/build/282899697
```
</details>

emilazy reviewed (approved):
Looks good – is there a reason not to default to Accelerate for Darwin though since apparently it’s what upstream defaults to and tests in CI? (And also maybe the last commit should be squashed into the first one given the different approach?)

paparodeo commented:
> Looks good – is there a reason not to default to Accelerate for Darwin though since apparently it’s what upstream defaults to and tests in CI? (And also maybe the last commit should be squashed into the first one given the different approach?)

I left openblas because it works on linux and that's what this was using before the SDK refactor and we've seen some sketchy Accelerate behavior so it seemed like a safer option.

tomberek commented:
Builds on aarch64-darwin.
